Add CI tests for `make distclean` and Porting/manicheck

nwc10 opened the pull request:
Also leave the checkout `fetch-depth` at its default (1) for most jobs.

nwc10 committed:
In manicheck be explicit that the code is stripping './'

The regex had been written as /^../, which strictly was correct as the
string passed would *always* start with './', but this wasn't clear.

One had to stop and double check how File::Find::find() works, and what
argument it was called with. And it's not exactly clear when the '.' in
the regex text matches '.' and then '/', but the '/' in the regex is the
delimiter.

Given that the regex is being changed, move it later - previously the code
was performing a substitution on the value of $File::Find::name before it
knew that it needed it. (ie doing work for all directories.)

nwc10 committed:
manicheck now optionally exits non-zero if it finds problems

This will enable it to be used in a CI test.

Refactor the code to avoid the temporary array @files by iterating over the
lines of MANIFEST as they are read in.

nwc10 committed:
Add CI tests for `make distclean` and Porting/manicheck

nwc10 committed:
Leave the checkout `fetch-depth` at its default (1) for most jobs

The default is a shallow checkout.

Keep the full checkout for the "sanity test" job so that it runs all the
tests in t/porting. Add a comment explaining the reasoning.

nwc10 commented:
@wolfsage - I wasn't sure whether some/all of the steps in the RMG about "check distclean" could now be removed, so didn't change the text. But as this gives us a CI test that will fail if `make distclean` and `perl Porting/manicheck` disagree, perhaps they can go?

jkeenan commented on Porting/manicheck:
```diff
@@ -4,28 +4,57 @@
+    my ($file) = $line =~ /^(\S+)/;
+    ++$files{$file};
+    next
+        if -f $file;
```
Why is the `next` statement broken into two lines?

nwc10 replied:
> Why is the `next` statement broken into two lines?

Because I was writing it in the style i prefer. Which is inconsistent with the rest of the file.

Oops, well spotted. Thanks for catching this.

Fixed in the rebase.

nwc10 committed:
A CI test that manicheck has no errors after `git clean -dxf`

nwc10 committed:
A CI test that manicheck reports no errors on a clean checkout

Do this on the "minitest" job as that has no "manicheck" step yet, and the
"minitest" job does not have a "matrix", so we only run this actual test
once.

wolfsage commented:
> @wolfsage - I wasn't sure whether some/all of the steps in the RMG about "check distclean" could now be removed, so didn't change the text. But as this gives us a CI test that will fail if `make distclean` and `perl Porting/manicheck` disagree, perhaps they can go?

For point releases, we don't (ever?) add any new files, so maybe. But what about new releases? (A lot of the releasing happens locally without pushing, and we don't push until after we've uploaded a tarball to cpan, so it could be too late?)

The second distclean check is for the built tarball, so maybe that one should stay regardless? I'm not sure

nwc10 commented:
> > @wolfsage - I wasn't sure whether some/all of the steps in the RMG about "check distclean" could now be removed, so didn't change the text. But as this gives us a CI test that will fail if `make distclean` and `perl Porting/manicheck` disagree, perhaps they can go?
> 
> For point releases, we don't (ever?) add any new files, so maybe. But what about new releases? (A lot of the releasing happens locally without pushing, and we don't push until after we've uploaded a tarball to cpan, so it could be too late?)

I doubt that we add new **installed** files to a maint point release, but we might add new test scripts as a consequence of backporting a commit that fixes a bug. (And if I can think of that case, there are probably other cases I can't think of)

> The second distclean check is for the built tarball, so maybe that one should stay regardless? I'm not sure

If it's for the built tarball, it's testing something we don't/can't test in a git checkout. "More" obviously it would fail if the tarball fails to include a file mentioned in `MANIFEST` (which might be possible without the build itself failing). I think we can build things differently if we're not in a git checkout, and that might mean we create different files, so we do need a check that our `distclean` rules also get these.

Ultimately, I think the answer is "don't remove the steps". It's useful to keep testing them, but the CI tests here mean that they shouldn't ever actually fail again, which saves future release managers the distraction that you got surprised with.

I think that this branch now better than what we had, so is good to merge. Unless someone (myself included) spots something daft I'll merge it if/when the CI run completes.
